### PR TITLE
Add Service Fabric service endpoint

### PR DESCRIPTION
**Pull request by hbuckle**
This PR adds a new resource for service endpoint for Service Fabric.

## All Submissions:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.
* [x] My code follows the code style of this project.
* [x] I ran lint checks locally prior to submission.
* [x] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?

Issue Number: #38

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [ ] Yes
- [x] No

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Any relevant logs, error output, etc?

(If it’s long, please paste to https://ghostbin.com/ and insert the link here.)

### CODE REVIEW

File: website/docs/r/serviceendpoint_servicefabric.html.markdown [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    server_certificate_thumbprint = "test"
+    username                      = "username"
+    password                      = "password"
+  }
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
```
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_servicefabric.html.markdown:89]: `cluster_endpoint`  was not documented

File: website/docs/r/serviceendpoint_servicefabric.html.markdown [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  - `client_certificate` - (Required) Base64 encoding of the cluster's client certificate file.
+  - `client_certificate_password` - (Required) Password for the certificate.
+- `azure_active_directory`
+  - `server_certificate_lookup` - (Required) Verification mode for the cluster. Possible values include `Thumbprint` or `CommonName`.
+  - `server_certificate_thumbprint` - (Optional) The thumbprint(s) of the cluster's certificate(s). This is used to verify the identity of the cluster. This value overrides the publish profile. Separate multiple thumbprints with a comma (',')
+  - `server_certificate_common_name` - (Optional) The common name(s) of the cluster's certificate(s). This is used to verify the identity of the cluster. This value overrides the publish profile. Separate multiple common names with a comma (',')
+  - `username` - (Required) - Specify an Azure Active Directory account.
+  - `password` - (Required) - Password for the Azure Active Directory account.
```
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_servicefabric.html.markdown]:
```suggestion


- `certificate` - (Optional) A `certificate` block as documented below.
- `azure_active_directory` - (Optional) A `azure_active_directory` block as documented below.

A `certificate`  block supports the following:
  - `client_certificate` - (Required) Base64 encoding of the cluster's client certificate file.
  - `client_certificate_password` - (Required) Password for the certificate.
  - `server_certificate_lookup` - (Required) Verification mode for the cluster. Possible values include `Thumbprint` or `CommonName`.
  - `server_certificate_thumbprint` - (Optional) The thumbprint(s) of the cluster's certificate(s). This is used to verify the identity of the cluster. This value overrides the publish profile. Separate multiple thumbprints with a comma (',')
  - `server_certificate_common_name` - (Optional) The common name(s) of the cluster's certificate(s). This is used to verify the identity of the cluster. This value overrides the publish profile. Separate multiple common names with a comma (',')

A  `azure_active_directory`  block supports the following:
  - `username` - (Required) - Specify an Azure Active Directory account.
  - `password` - (Required) - Password for the Azure Active Directory account.
  - `server_certificate_lookup` - (Required) Verification mode for the cluster. Possible values include `Thumbprint` or `CommonName`.
  - `server_certificate_thumbprint` - (Optional) The thumbprint(s) of the cluster's certificate(s). This is used to verify the identity of the cluster. This value overrides the publish profile. Separate multiple thumbprints with a comma (',')
  - `server_certificate_common_name` - (Optional) The common name(s) of the cluster's certificate(s). This is used to verify the identity of the cluster. This value overrides the publish profile. Separate multiple common names with a comma (',')

```

File: website/docs/r/serviceendpoint_servicefabric.html.markdown [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+resource "azuredevops_serviceendpoint_servicefabric" "test" {
+  project_id            = data.azuredevops_project.p.id
+  service_endpoint_name = "test"
+  description           = "test"
+  cluster_endpoint      = "tcp://test"
+
+  certificate {
+    server_certificate_lookup     = "Thumbprint"
+    server_certificate_thumbprint = "test"
+    client_certificate            = filebase64("${path.module}/certificate.pfx")
```
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_servicefabric.html.markdown]: Suggest change to 
```suggestion
    client_certificate            = filebase64("certificate.pfx")
```

File: website/docs/r/serviceendpoint_servicefabric.html.markdown [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+resource "azuredevops_serviceendpoint_servicefabric" "test" {
+  project_id            = data.azuredevops_project.p.id
+  service_endpoint_name = "test"
+  description           = "test"
+  cluster_endpoint      = "tcp://test"
+
+  azure_active_directory {
+    server_certificate_lookup     = "Thumbprint"
+    server_certificate_thumbprint = "test"
```
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_servicefabric.html.markdown]:
```suggestion
    server_certificate_thumbprint = "0000000000000000000000000000000000000000"
```

File: website/docs/r/serviceendpoint_servicefabric.html.markdown [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+```hcl
+data "azuredevops_project" "p" {
+  name = "contoso"
+}
+
+resource "azuredevops_serviceendpoint_servicefabric" "test" {
+  project_id            = data.azuredevops_project.p.id
+  service_endpoint_name = "test"
+  description           = "test"
```
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_servicefabric.html.markdown]:
```suggestion
  service_endpoint_name = "Sample Service Fabric"
  description           = "Managed by Terraform"
```

File: website/docs/r/serviceendpoint_servicefabric.html.markdown [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+Manages a Service Fabric service endpoint within Azure DevOps.
+
+## Example Usage
+
+### Client Certificate Authentication
+
+```hcl
+data "azuredevops_project" "p" {
+  name = "contoso"
+}
```
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_servicefabric.html.markdown]:
```suggestion
resource "azuredevops_project" "p" {
  name       = "Test Project"
  description        = "Test Project Description"
  visibility         = "private"
  version_control    = "Git"
  work_item_template = "Agile"

  features = {
      "testplans" = "disabled"
      "artifacts" = "disabled"
  }
}
```

File: website/docs/r/serviceendpoint_servicefabric.html.markdown [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+  }
+}
+```
+
+### Azure Active Directory Authentication
+
+```hcl
+data "azuredevops_project" "p" {
+  name = "contoso"
+}
```
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_servicefabric.html.markdown]:
```suggestion
resource "azuredevops_project" "p" {
  name       = "Test Project"
  description        = "Test Project Description"
  visibility         = "private"
  version_control    = "Git"
  work_item_template = "Agile"
  features = {
      "testplans" = "disabled"
      "artifacts" = "disabled"
  }
}
```

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_servicefabric.go [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+		result[0]["server_certificate_common_name"] = (*serviceEndpoint.Authorization.Parameters)["servercertcommonname"]
+	}
+	return result
+}
+
+// Convert AzDO data structure to internal Terraform data structure
+func flattenServiceEndpointServiceFabric(d *schema.ResourceData, serviceEndpoint *serviceendpoint.ServiceEndpoint, projectID *string) {
+	doBaseFlattening(d, serviceEndpoint, projectID)
+
+	switch *serviceEndpoint.Authorization.Scheme {
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_servicefabric.go:260]: Service fabric support three type schema: `Certificate`, `UsernamePassword` and `None`. Current resource does not support `None`. If someone import an service fabric create in schema `None`,  provider should  notify user that this resource was not supported.

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_servicefabric_test.go [hunk 9]
```diff
@@ -0,0 +1,166 @@
+// +build all resource_serviceendpoint_github
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_servicefabric_test.go]:
```suggestion
// +build all resource_serviceendpoint_service_faric
```

File: azuredevops/internal/acceptancetests/resource_serviceendpoint_servicefabric_test.go [hunk 10]
```diff
@@ -0,0 +1,59 @@
+// +build all resource_serviceendpoint_service_fabric
+// +build !exclude_serviceendpoints
+
+package acceptancetests
+
+import (
```
Comment (xuzhang3) [inline azuredevops/internal/acceptancetests/resource_serviceendpoint_servicefabric_test.go:6]: `azure_active_directory` not covered by AccTests. 
Tests should cover different  scenario like, `basic create`, `update`, `complete  create`

**Review by xuzhang3 — CHANGES_REQUESTED** [review]

**Comment by hbuckle** [comment]
@xuzhang3 - the `none` authorisation type has now been added

File: website/docs/r/serviceendpoint_servicefabric.html.markdown [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+- `project_id` - (Required) The project ID or project name.
+- `  service_endpoint_name = "Sample Service Fabric"
+  description           = "Managed by Terraform"faults to `Managed by Terraform`.
+- `cluster_endpoint` - (Required) Client connection endpoint for the cluster. Prefix the value with 'tcp://';. This value overrides the publish profile.
```
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_servicefabric.html.markdown]:
```suggestion
- `project_id` - (Required) The project ID or project name.
- `service_endpoint_name` - (Required) The Service Endpoint name.
- `cluster_endpoint` - (Required) Client connection endpoint for the cluster. Prefix the value with 'tcp://';. This value overrides the publish profile.
- `description` - (Optional) The Service Endpoint description. Defaults to `Managed by Terraform`.

```

File: website/docs/r/serviceendpoint_servicefabric.html.markdown [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+## Relevant Links
+
+- [Azure DevOps Service REST API 5.1 - Endpoints](https://docs.microsoft.com/en-us/rest/api/azure/devops/serviceendpoint/endpoints?view=azure-devops-rest-5.1)
+
+## Import
+
+Azure DevOps Service Endpoint Service Fabric can be imported using **projectID/serviceEndpointID** or **projectName/serviceEndpointID**
+
+```sh
+ terraform import azuredevops_serviceendpoint_servicefabric.serviceendpoint 00000000-0000-0000-0000-000000000000/00000000-0000-0000-0000-000000000000
```
Comment (xuzhang3) [inline website/docs/r/serviceendpoint_servicefabric.html.markdown]:
```suggestion
$ terraform import azuredevops_serviceendpoint_servicefabric.serviceendpoint 00000000-0000-0000-0000-000000000000/00000000-0000-0000-0000-000000000000

```

File: azuredevops/internal/acceptancetests/resource_serviceendpoint_servicefabric_test.go [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+// validates that an apply followed by another apply (i.e., resource update) will be reflected in AzDO and the
+// underlying terraform state.
+func TestAccServiceEndpointServiceFabric_CertificateCreateAndUpdate(t *testing.T) {
+	projectName := testutils.GenerateResourceName()
+	serviceEndpointNameFirst := testutils.GenerateResourceName()
+	serviceEndpointNameSecond := testutils.GenerateResourceName()
+
+	resourceType := "azuredevops_serviceendpoint_servicefabric"
+	tfSvcEpNode := resourceType + ".serviceendpoint"
+	resource.Test(t, resource.TestCase{
```
Comment (xuzhang3) [inline azuredevops/internal/acceptancetests/resource_serviceendpoint_servicefabric_test.go]: Suggest change to `resource.ParallelTest`

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_servicefabric.go [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+				"client_certificate": {
+					Type:             schema.TypeString,
+					Required:         true,
+					Description:      "Base64 encoding of the cluster's client certificate file.",
+					Sensitive:        true,
+					DiffSuppressFunc: tfhelper.DiffFuncSuppressSecretChanged,
+				},
+				"client_certificate_password": {
+					Type:             schema.TypeString,
+					Required:         true,
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_servicefabric.go]: `client_certificate_password` is not a required configuration

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_servicefabric.go [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+		MaxItems: 1,
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"server_certificate_lookup":      servicefabricServerCertificateLookupSchema(),
+				"server_certificate_thumbprint":  servicefabricServerCertificateThumbprintSchema(resourceBlockServiceFabricAzureActiveDirectory),
+				"server_certificate_common_name": servicefabricServerCertificateCommonNameSchema(resourceBlockServiceFabricAzureActiveDirectory),
+				"username": {
+					Type:        schema.TypeString,
+					Required:    true,
+					Description: "Specify an Azure Active Directory account.",
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_servicefabric.go]: Should add a valitaion: `ValidateFunc: validation.StringIsNotEmpty,`  for `username`

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_servicefabric.go [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+		}, false),
+	}
+}
+
+func servicefabricServerCertificateThumbprintSchema(blockName string) *schema.Schema {
+	return &schema.Schema{
+		Type:          schema.TypeString,
+		Optional:      true,
+		Description:   "The thumbprint(s) of the cluster's certificate(s). This is used to verify the identity of the cluster. This value overrides the publish profile. Separate multiple thumbprints with a comma (',')",
+		ConflictsWith: []string{fmt.Sprintf("%s.0.server_certificate_common_name", blockName)},
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_servicefabric.go:137]: Should add a validation: `ValidateFunc: validation.StringIsNotEmpty,`

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_servicefabric.go [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+	}
+}
+
+func servicefabricServerCertificateCommonNameSchema(blockName string) *schema.Schema {
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_servicefabric.go:141]: Should add a validation: `ValidateFunc: validation.StringIsNotEmpty,`

**Review by xuzhang3 — CHANGES_REQUESTED** [review]

File: azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_servicefabric.go [hunk 18]
```diff
@@ -0,0 +1,276 @@
...
+				"unsecured": {
+					Type:        schema.TypeBool,
+					Optional:    true,
+					Default:     false,
+					Description: "Skip using windows security for authentication.",
+				},
+				"cluster_spn": {
+					Type:         schema.TypeString,
+					Optional:     true,
+					ValidateFunc: validation.StringIsNotEmpty,
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/resource_serviceendpoint_servicefabric.go]: `cluster_spn` accept empty string, Use validation like  `validation.StringLenBetween(0, 1024),` check the value length is the better choice.

**Review by xuzhang3 — CHANGES_REQUESTED** [review]

**Comment by xuzhang3** [comment]
LGTM

**Pull request merged**